Simplenote/src/main/java/com/automattic/simplenote/NoteListFragment.java

```diff
@@ -740,10 +761,71 @@ public boolean hasSearchQuery() {
         return mSearchString != null && !mSearchString.equals("");
     }
 
+    public void addSearchItem(String item, int index) {
+        List<String> recents = new ArrayList<>();
+        Query<Preferences> query = mBucketPreferences.query();
+
+        try (ObjectCursor<Preferences> cursor = query.execute()) {
+            Preferences preferences = mBucketPreferences.get(PREFERENCES_OBJECT_KEY);
+
+            if (cursor.moveToNext()) {
+                recents = cursor.getObject().getRecentSearches();
+            }
+
+            recents.remove(item);
+            recents.add(index, item);
+            // Trim recent searches to MAX_RECENT_SEARCHES (currently 5) if size is greater than MAX_RECENT_SEARCHES.
+            preferences.setRecentSearches(recents.subList(0, recents.size() > MAX_RECENT_SEARCHES ? MAX_RECENT_SEARCHES : recents.size()));
+            preferences.save();
+        } catch (BucketObjectMissingException exception) {
+            // Create entity for recent searches since it doesn't exist.
+            try {
+                Preferences preferences = mBucketPreferences.newObject(PREFERENCES_OBJECT_KEY);
+                preferences.save();
+            } catch (BucketObjectNameInvalid invalid) {
+                Log.e("addSearchItem", "Could not create search object", invalid);
+            }
```

Comment:
It seems like this function and some others would be much clearer if we didn't have to expose the logic to grab the preferences entity _and_ update it. The extra code noise at least makes it hard for me to understand what is essential here vs. what is incidental…

```java
public Preferences getPreferences() {
    try {
        return mBucketPreferences.get(PREFERENCES_OBJECT_KEY);
    } catch (BucketObjectMissingException e) {
        // oops, we don't have the preferences entity yet
    }

    try {
        Preferences preferences = mBucketPreferences.newObject(PREFERENCES_OBJECT_KEY);
        preferences.save();
        return preferences;
    } catch (BucketObjectNameInvalid invalid) {
        Log.e("getPreferences", "Could not create preferences entity", invalid);
        return null; // nothing we can do
    }
}

public void addSearchItem(String item, int index) {
    Preferences preferences = getPreferences();
    if (null == preferences) return;
    List<String> recentSearches = preferences.getRecentSearches();

    recentSearches.remove(item);
    recentSearches.add(index, item);

    // Trim recent searches to MAX_RECENT_SEARCHES (currently 5) if size is greater than MAX_RECENT_SEARCHES.
    preferences.setRecentSearches(recentSearches.subList(0, Math.min(MAX_RECENT_SEARCHES, recentSearches.size())));
    preferences.save();
}
```

When I tested this it seemed to work as expected. I tried hard to understand what we're doing with the `query` and `cursor` but in the end could not. It seems like the `query` iterates over all the entities in the bucket (which in this case should be only one) and the cursor jumps to the first one.

The thing is, we grab the preferences when we do `mNoteBucket.get(PREFERENCES_OBJECT_KEY)` and this should always be the same as what the query and cursor give us - if they were any different at all we'd be in a very broken place.

So, is the query necessary at all? Why not just use `preferences.getRecentSearches()`?

I saw this pattern somewhere else - do you find the separation of getting the preferences entity helpful? I suspect that it's all complicated by trying to treat the bucket as an entity and not a major deal, just confusing.